routers/web/repo/pull.go

```diff
@@ -395,12 +395,17 @@ func PrepareViewPullInfo(ctx *context.Context, issue *issues_model.Issue) *git.C
 	var headBranchSha string
 	// HeadRepo may be missing
 	if pull.HeadRepo != nil {
-		headGitRepo, err := gitrepo.OpenRepository(ctx, pull.HeadRepo)
-		if err != nil {
-			ctx.ServerError("OpenRepository", err)
-			return nil
+		var headGitRepo *git.Repository
+		if ctx.Repo != nil && ctx.Repo.Repository != nil && pull.HeadRepoID == ctx.Repo.Repository.ID && ctx.Repo.GitRepo != nil {
+			headGitRepo = ctx.Repo.GitRepo
+		} else {
+			headGitRepo, err = gitrepo.OpenRepository(ctx, pull.HeadRepo)
```

Comment:
IIRC `OpenRepository` is quite lightweight, it only returns a new struct pointer.

Does it really need that complex `if` checks?

Reply:
As of https://github.com/go-gitea/gitea/pull/31754, this just checks whether or not the folder exists, which is some disk IO we can avoid. 

However, other operations will open a new `cat-file --batch`, which will be reused against the same object.

Reply:
> As of #31754, this just checks whether or not the folder exists, which is some disk IO we can avoid.

It seems quite trivial.

> However, other operations will open a new `cat-file --batch`, which will be reused against the same object.

In this case, does it really reuse? If a "CatBatch" is already in use, the same repo still opens new "CatBatch", it can't be shared.

----

So still the question, does it really need that complex if checks? I would prefer to keep the logic simple if the benefit is not significant

Reply:
This is the same pattern used throughout this file. For example, see https://github.com/go-gitea/gitea/blob/7f21b3fa6c811e9dd531db10fc423ed2b647b253/routers/web/repo/pull.go#L1207. It appears that this specific instance was just missed.

Reply:
Well, TBH I do not think the old code is right. 😅

Reply:
IIRC there is a `RepositoryFromContextOrOpen`, which was designed for this case.